synapse/handlers/event_auth.py

```diff
@@ -59,32 +59,55 @@ async def check_restricted_join_rules(
         ):
             return
 
+        # Get the spaces which allow access to this room, if applicable.
+        allowed_spaces = await self.get_spaces_that_allow_join(state_ids, room_version)
+        if allowed_spaces is None:
+            return
+
+        # The user was not in any of the required spaces.
+        if not await self.is_user_in_rooms(allowed_spaces, user_id):
+            raise AuthError(
+                403,
+                "You do not belong to any of the required spaces to join this room.",
+            )
+
+    async def get_spaces_that_allow_join(
+        self, state_ids: StateMap[str], room_version: RoomVersion
+    ) -> Optional[Collection[str]]:
+        """
+        Generate a list of spaces which allow access to a room.
+
+        Args:
+            state_ids: The state of the room as it currently is.
+            room_version: The room version of the room to query.
+
+        Returns:
+            None if room access via spaces doesn't apply for this room.
+
+            Otherwise, a collection of spaces (if any) which provide membership
+            to the room.
+        """
         # This only applies to room versions which support the new join rule.
         if not room_version.msc3083_join_rules:
-            return
+            return None
 
         # If there's no join rule, then it defaults to invite (so this doesn't apply).
         join_rules_event_id = state_ids.get((EventTypes.JoinRules, ""), None)
         if not join_rules_event_id:
-            return
+            return None
 
         # If the join rule is not restricted, this doesn't apply.
         join_rules_event = await self._store.get_event(join_rules_event_id)
         if join_rules_event.content.get("join_rule") != JoinRules.MSC3083_RESTRICTED:
-            return
+            return None
```

Comment:
I'm wondering if splitting this out into `has_restricted_join_rules` would make sense? That way `get_spaces_that_allow_join` can always return a list and callers that care can do the explicit check? I think the only place the difference matters is `check_restricted_join_rules`, where being explicit probably helps? IDK.

Reply:
All the callers care if the join rules are active or not? But some need to differentiate between whether they are activate vs an empty list? Is that what you're getting at? I'll try it and see how it looks.

Reply:
I'm not sure if this helps at all, the 4 callers end up just calling another method before it. It might be a bit clearer though about what's happening.

I pushed d7206152b6be446981beae4e83f863d9973d1294 with the change though.